source/common/http/conn_manager_utility.cc

```diff
@@ -90,7 +90,17 @@ ConnectionManagerUtility::MutateRequestHeadersResult ConnectionManagerUtility::m
   // If this is a Upgrade request, do not remove the Connection and Upgrade headers,
   // as we forward them verbatim to the upstream hosts.
   if (!Utility::isUpgrade(request_headers)) {
-    request_headers.removeConnection();
+    // Per RFC2616, connection keepalive is optional and useful for java/tomcat.
+    if (protocol == Protocol::Http11 && request_headers.Connection() &&
+        Envoy::StringUtil::caseFindToken(request_headers.Connection()->value().getStringView(), ",",
+                                         Http::Headers::get().ConnectionValues.KeepAlive)) {
```

Comment:
I don't think this is the route we want to go. Upstream keep-alive and keep-alive duration should be seprate from downstream keep-alive.  I think it'd be fine to add a keep alive to the response sent downstream but I'm unsure why we'd want to proxy one through.  Can you be a bit more clear about why you want the same timeout persisted?
If you go this route it 100% has to be a long term config option and be added to HCM API surface, as I don't think most folks will want it so reloadble flags don't seem optimal.

Reply:
Thanks  for review.  I think sending keep-alive from upstream response header is neat. Issue is different servers have different behavior e.g. if upstream is java/tomcat based, it will not emit keep-alive timeout settings in response  header by default until client sends `connection:keep-alive` which jdk clients does  but it gets dropped by proxy. We are not honoring downstream keep-alive time out in request header, just connection header. PTAL at ref test doc for nodejs and java/tomcat behind proxy https://docs.google.com/document/d/1JNynMtGAy5czphdJzvutgoVKG3UGqfo3CMVheea6O-E/edit . I am ok to skip `connection:keep-alive `in request header if can call out that customers using apache/tomcat behind proxy/envy, tweaksdirectly on the server side to send keep-alive response header by default similar to how nodejs does. As long as upstream emits, envoy can honor that part of this change. 

May be long term sure can explore to convert to HCP API vs runtime guard for now.

Reply:
I'm still unclear why we want to use the same keep-alive.  The issue you link (https://github.com/envoyproxy/envoy/issues/17613) involves Envoy setting its own keep-alive header to make it clear to clients when Envoy will close the connection, which seems reasonable.  Envoy forwarding an upstream keep-alive timeout which doesn't correspond to any Envoy timeout configuration continues to make no sense to me.   The network set up here is upstream ---- Envoy ---- client.  The upstream to Envoy connection has a completely different lifetime than the Envoy to client connection, and unless you can give a compelling reason to forward one connection's timeout to another unrelated connection, I don't think this change should proceed.  If you do have a compelling reason, it is not OK for this to land as a runtime guarded change, as it violates Envoy data plane principles and would likely cause harm to users.  It needs to be a config guarded option if it lands at all.

Reply:
Thanks.Seems with hardware load balancers, it forwards the same server keep-alive as is until we explicitly over-ride .Hence, kept the behavior intact vs we can also add some default keep-alive capability based on say idletimeout settings, etc. but this seems more neat to align what backend/upstream has in its default. Because after move to proxy based LB, we see ssl connection creation spikes up as keep-alive timeout itself is missing causing clients to create new ssl connections for each requests vs reuse causing more CPU load on envoy. So I am ok to change timeout values later too but original intent to pass the in response header still stays same. Let me know further 

So do you want me to make this as a config guard vs runtime in this patch or an incremental change is ok too? Please let me know.

Reply:
@alyssawilk hello just following up on this again. I dialed in the community meeting and seems there wasnt any audience so wanted to have a gentle reminder as we use this patch internally and helps overall. Ok to do it incrementally by returning upstream default keep-alive for h1.1 and have an add on to send envoy's default. Reason is returning just envoy's default will not help if upstream/server is not honoring it for e2e flow. Let me know cc @mattklein123

Reply:
Hardware load balancers may actually tie the lifetime of the two connections, Envoy does not so that argument still isn't particularly compelling.  Meanwhile this PR is marked waiting and will not get further review until you addressed the comments in my last pass, which you have not done.

Reply:
Thanks a bunch. I was waiting to clarify. So we we ok with this patch but address it with config changes vs runtime guard? Please advice and I will rectify accordingly to make it config guard change vs runtime guard as we use it internally via runtime guard and works fine too. can update the ssl metrics and cpu improvements results in the same doc. Let me know.

Reply:
@noah8713 It looks like we should add something like `connection_headers_to_preserve: ['keep-alive']` to [HttpConnectionManager](https://github.com/envoyproxy/envoy/blob/152cf3d30319fc739cd84e47f3147eec6d1b54b4/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L41) proto.  Though this behavior would violate the connection-token removal requirements of https://datatracker.ietf.org/doc/html/rfc2616#section-8.1.3, we have definitely seen material improvements in the controlled environment of our service mesh.

Reply:
Thanks @alyssawilk for confirming. @kb000 I think it would be good to expose similar to case preserve use case via HCM under http_protocol_options where we have [idle timeouts and other bits present](https://github.com/envoyproxy/envoy/blob/main/api/envoy/api/v2/core/protocol.proto#L72). Can move retain_keepalive_header_http11 option under the same and see how it goes.

Reply:
Another thing I completely missed is runtime guard is supported as an option from istio via [runtimevalues](https://github.com/istio/istio/blob/master/pkg/bootstrap/config.go#L320) so reverting the HCM config will incur further changes to honor this either via envoyfilter CRD or via some additional changes in istio. I see there are too many guards for http bits e.g. skip_delay_close, etc. So can we keep it if so? I should have called out earlier. Let me know if we can keep runtime guard for first cut as its not a major change too and works fine in our service mesh with no additional changes in istio side. If its a big no, I will proceed to make HCM side change. Please confirm,

Reply:
Thanks for suggestions. I moved this as a config under HCM. PTAL @alyssawilk

Reply:
As said I'm going to defer to the Envoy API shephard, @mattklein123